File: hejoo345/week_16/PGS_단어 변환.js
function solution(begin, target, words) {
    var answer = 0;
    let queue = [];
    let visited = []

    if (!words.includes(target)) { // words에 target이 없으면 0 반환
        return 0;
    }

    queue.push([begin, answer]);
    visited.push(begin);

    while (queue) {
        const [word, cnt] = queue.shift();
        if (word == target) return cnt; // target과 일치하면 카운트 반환

        words.forEach(item => {
            let notEq = 0;
            if (visited.includes(item)) return; // forEach 종료
            for (let i = 0; i < item.length; i++) {
                if (word[i] != item[i]) {
                    notEq++;
                }
            }
            if (notEq == 1) {
                queue.push([item, cnt + 1]);
                visited.push(item);
            }
        })
    }
    return answer;
}

File: OneK-2/week_16/PGS_단어 변환.java
class Solution {
    static boolean[] visited;
    static int answer = 0;

    public int solution(String begin, String target, String[] words) {
        visited = new boolean[words.length];
        dfs(begin, target, words, 0);

        return answer;
    }

    private static void dfs(String cur, String target, String[] words, int cnt) {
        if (cur.equals(target)) {
            answer = cnt;
            return;
        }

        for (int i = 0; i < words.length; i++) {
            if (visited[i]) continue;
            int num = 0;
            for (int j = 0; j < cur.length(); j++) {
                if (cur.charAt(j) != words[i].charAt(j)) {
                    num++;
                }
            }
            if (num == 1) {
                visited[i] = true;
                dfs(words[i], target, words, cnt + 1);
                visited[i] = false;
            }
        }
    }
}